configure.ac

```diff
@@ -219,24 +219,39 @@ AC_CHECK_LIB([gsl],[gsl_sf_bessel_Jn], [],
 # Check for HDF5 library
 AC_ARG_WITH(hdf5, [AC_HELP_STRING([--without-hdf5],[do not allow HDF5 output])], ok=$withval, ok=yes)
 if test "$ok" = "yes"; then
-	AC_CHECK_LIB(z,deflate, [],[AC_MSG_WARN([zlib is required for HDF5!])])
-
-	save_CC=$CC
-	if test "x$with_mpi" = "xyes"; then
-	   CC=$MPICXX
-	fi
-	save_LIBS_0="$LIBS" # save, to check later if we found any library
-	AC_CHECK_LIB(hdf5, H5Pcreate, [
-		AC_CHECK_HEADERS(hdf5.h, [LIBS="-lhdf5 $LIBS"
-                     AC_DEFINE(HAVE_HDF5,1,[Define if we have & link HDF5])])])
-	if test x"$save_LIBS_0" = x"$LIBS"; then
-		AC_MSG_WARN([Couldn't find the HDF5 library!!  Switching to --without-hdf5.])
-	fi
-	CC=$save_CC
-
-        if test "x$with_mpi" = "xyes"; then
-		AC_CHECK_FUNCS(H5Pset_mpi H5Pset_fapl_mpio)
-	fi
+    AC_CHECK_LIB(z,deflate, [],[AC_MSG_WARN([zlib is required for HDF5!])])
+
+    save_CC="$CC"
+    save_CPPFLAGS="$CPPFLAGS"
+    save_LIBS="$LIBS"
+
+    # Attempt to locate HDF5 via pkg-config and fall back to a default.
+    # Optimistically update LIBS and CPPFLAGS. Revert if any AC_CHECKs fail.
+    PKG_CHECK_MODULES(HDF5, hdf5, [
```

Comment:
One issue with using `PKG_CHECK_MODULES` like this is that if the user explicitly set `-L` and `-I` flags to find a particular version of HDF5 on their system, using `pkg-config` unconditionally will override that choice.  It is probably better to do `AC_CHECK_LIB` first.   See also https://stackoverflow.com/questions/10220946/pkg-check-modules-considered-harmful

In particular, it seems like we should so something like
```sh
# HDF5 libraries may need to be linked with mpicxx if they were compiled with MPI
if test "x$with_mpi" = "xyes"; then
    CC=$MPICXX
fi

# see if -lhdf5 works, then fall back to pkg-config flags (checked with AC_CHECK_FUNC below)
AC_CHECK_LIB(hdf5, H5Pcreate, [], [
    PKG_CHECK_MODULES(HDF5, hdf5, [
        CPPFLAGS="$CPPFLAGS $HDF5_CFLAGS" # assumes that CFLAGS actually contains only -I flags
        LIBS="$HDF5_LIBS $LIBS"
    ], [ac_cv_header_hdf5_h=no])
])
```
and then change the test below to `AC_CHECK_FUNC` instead of `AC_CHECK_LIBS`.   We can also just run `AC_CHECK_FUNCS(H5Pset_mpi H5Pset_fapl_mpio)` unconditionally below.

Reply:
I updated the PR based on your suggestions:
Check whether `-lhdf5` just works, otherwise try `pkg-config`.
Use AC_CHECK_FUNC and AC_CHECK_HEADER to determine whether the derived configuration works correctly.
Unconditionally check for `H5Pset_mpi` and `H5Pset_fapl_mpio`.